source/op/prod_env_mat_multi_device.cc

```diff
@@ -68,7 +69,30 @@ descrpt: The environment matrix.
 descrpt_deriv: The derivative of the environment matrix.
 rij: The distance between the atoms.
 nlist: The neighbor list of each atom.)");
-    // only sel_a and rcut_r used.
+// only sel_a and rcut_r used.
+
+#else
+// alias of ProdEnvMatAMesh, only used in Ascend device inference.
+REGISTER_OP("ProdEnvMatAMesh")
+    .Attr("T: {float, double} = DT_DOUBLE")
+    .Input("coord: T")          //atomic coordinates
+    .Input("type: int32")       //atomic type
+    .Input("natoms: int32")     //local atomic number; each type atomic number; daizheyingxiangqude atomic numbers
+    .Input("box : T")
+    .Input("mesh : int32")
+    .Input("davg: T")           //average value of data
+    .Input("dstd: T")           //standard deviation
+    .Attr("rcut_a: float")      //no use
+    .Attr("rcut_r: float")
+    .Attr("rcut_r_smth: float")
+    .Attr("sel_a: list(int)")
+    .Attr("sel_r: list(int)")   //all zero
+    .Output("descrpt: T")
+    .Output("descrpt_deriv: T")
+    .Output("rij: T")
+    .Output("nlist: int32");
+    // only sel_a and rcut_r uesd.
```

Comment:
It is unclear to me why a new OP is created. The parameters look the same as `ProdEnvMatA`.

Reply:
The ProdEnvMatA op in the Ascend package is incomplete, and it can not build the neighbor list, that will not affect the lammps inference. Since dp test inference needs to build a neighbor table in this op, we repackage the building neighbor list part in this op, which runs on the CPU. This op is only called in DP test to generate the neighbor list.

Reply:
If they have the same inputs and the same outputs but just with different implementation, I would not suggest defining two OPs, but defining only one OP with different kernels. (`REGISTER_KERNEL_BUILDER`) So the same graph can be run across different devices.

Reply:
Thanks. We refactored the OP kernels, actually the output is different with ProdEnvMatA, so we have to define a new op.

Reply:
Thanks. We refactored the OP kernels, actually the output is different with ProdEnvMatA, so we have to define a new op.